### PR TITLE
Used to identify the Automatic Tank Gauge  protocol

### PR DESCRIPTION
**Pull request by wssxsxxsx**
Please sign (check) the below before submitting the Pull Request:

- [x] I have signed the ntop Contributor License Agreement at https://github.com/ntop/legal/blob/main/individual-contributor-licence-agreement.md
- [x] I have read the contributing guidelines at https://github.com/ntop/nDPI/blob/dev/CONTRIBUTING.md
- [x] I have updated the documentation (in doc/) to reflect the changes made (if applicable)

Link to the related [issue](https://github.com/ntop/nDPI/issues):

Describe changes:

### CODE REVIEW

File: src/lib/protocols/atg.c [hunk 1]
```diff
@@ -0,0 +1,70 @@
+/*
+ * zmq.c
```
Comment (utoni) [inline src/lib/protocols/atg.c]:
```suggestion
 * atg.c
```

File: src/lib/protocols/atg.c [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+ *
+ * Copyright (C) 2016-22 - ntop.org
```
Comment (utoni) [inline src/lib/protocols/atg.c]:
```suggestion
 * Copyright (C) 2024 - ntop.org
```

File: src/lib/protocols/viber.c [hunk 3]
```diff
@@ -42,14 +42,6 @@ static void ndpi_search_viber(struct ndpi_detection_module_struct *ndpi_struct,
...
   NDPI_LOG_DBG(ndpi_struct, "search for Viber\n");
 
-  if(packet->udp && packet->iph) {
-    /* ignore broadcast as this isn't viber */
-    if((packet->iph->saddr == 0xFFFFFFFF) || (packet->iph->daddr == 0xFFFFFFFF)) {
-      NDPI_EXCLUDE_PROTO(ndpi_struct, flow);
-      return;
-    }
-  }
-
```
Comment (utoni) [inline src/lib/protocols/viber.c]: Can you explain why removing this block?

File: src/lib/protocols/atg.c [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+ *
+ * nDPI is free software: you can zmqtribute it and/or modify
```
Comment (utoni) [inline src/lib/protocols/atg.c]: zmqtribute?

File: src/lib/protocols/atg.c [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    }
+  }
+  
+  NDPI_EXCLUDE_PROTO(ndpi_struct, flow); /* No luck this time */
+}
+  
+
+void init_atg_dissector(struct ndpi_detection_module_struct *ndpi_struct,
+				u_int32_t *id) {
+  ndpi_set_bitmask_protocol_detection("EthernetIP", ndpi_struct, *id,
```
Comment (utoni) [inline src/lib/protocols/atg.c]:
```suggestion
  ndpi_set_bitmask_protocol_detection("atg", ndpi_struct, *id,
```

File: src/lib/protocols/atg.c [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+
+static void ndpi_search_atg(struct ndpi_detection_module_struct *ndpi_struct,
+				    struct ndpi_flow_struct *flow) {
+  struct ndpi_packet_struct *packet = &ndpi_struct->packet;
+  
+  NDPI_LOG_DBG(ndpi_struct, "search for ATG\n");
+
+  if(packet->tcp != NULL) {
+    NDPI_LOG_DBG2(ndpi_struct, "calculating ATG over tcp\n");
```
Comment (utoni) [inline src/lib/protocols/atg.c]:
```suggestion
```

**Review by utoni — CHANGES_REQUESTED** [review]
Thanks for the dissector.
Can you please add a PCAP file related to `ATG` to `tests/cfgs/default/pcap/` and run `tests/do.sh` ?

File: src/lib/protocols/viber.c (hunk 3 above)
Comment (wssxsxxsx) [reply src/lib/protocols/viber.c]: It should have been deleted by mistake

File: src/lib/protocols/atg.c (hunk 1 above)
Comment (wssxsxxsx) [reply src/lib/protocols/atg.c]: ok，I'll upload it later soon

File: tests/cfgs/default/pcap/atg.pcap
Comment (utoni) [inline tests/cfgs/default/pcap/atg.pcap:1]: You need to run `./tests/do.sh` and `git add ./tests/cfgs/default/result/atg.pcap.out`.

**Comment by koltiradw** [comment]
I think this dissector does not fully cover this protocol. I think you need to add commands from this [article](https://nationalpetroleum.net/docs/veeder-root/tls450-serialcommands-manual.pdf). And if you pay attention to the request format, it turns out that the request starts with 0x1, then comes the command and at the end 0xd, 0xa, i.e. /r/n.

**Comment by wssxsxxsx** [comment]
I tried two identification methods and found that some atg packets did not have ending characters.

**Comment by wssxsxxsx** [comment]
> Thanks for the dissector. Can you please add a PCAP file related to `ATG` to `tests/cfgs/default/pcap/` and run `tests/do.sh` ?

I've added.Thank you for your guidance！

**Comment by utoni** [comment]
image

It does not state anything about `\r\n`. Any clues, @koltiradw?

**Comment by wssxsxxsx** [comment]
> Thanks for the dissector. Can you please add a PCAP file related to `ATG` to `tests/cfgs/default/pcap/` and run `tests/do.sh` ?

I've uploaded the result,。how do I resolve this request， Or I'll just wait ？（this is my first time submitting a PR）

**Comment by utoni** [comment]
Let's see if CI agrees ;)

**Comment by koltiradw** [comment]
@utoni, @wssxsxxsx 
I meant the contents of the pcap file:
image
image
image
image

File: src/lib/protocols/atg.c [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+static void ndpi_search_atg(struct ndpi_detection_module_struct *ndpi_struct,
+				    struct ndpi_flow_struct *flow) {
+  struct ndpi_packet_struct *packet = &ndpi_struct->packet;
+  
+  NDPI_LOG_DBG(ndpi_struct, "search for ATG\n");
+
+  if(packet->tcp != NULL) {
+    if(packet->payload_packet_len >= 8) {
+      u_int16_t eth_ip_port = ntohs(10001);
```
Comment (utoni) [inline src/lib/protocols/atg.c]: Is the port fixed for ATG?

**Review by utoni — CHANGES_REQUESTED** [review]
Please add `<ClCompile Include="..\src\lib\protocols\atg.c" />` to `windows/nDPI.vcxproj`.

**Comment by utoni** [comment]
@wssxsxxsx 
It would be better to implement the protocol dissector as described in the article that @koltiradw posted with `\r\n` support.

File: src/lib/protocols/atg.c [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+      if((packet->tcp->source == eth_ip_port) || (packet->tcp->dest == eth_ip_port)) {
+	if(packet->payload[0] == 0x01 ) {
```
Comment (utoni) [inline src/lib/protocols/atg.c]: Having just these detection pattern i.e. w/o the port check is not enough and will lead to lot's of false positives.

**Review by utoni — CHANGES_REQUESTED** [review]

File: src/lib/protocols/atg.c (hunk 7 above)
Comment (wssxsxxsx) [reply src/lib/protocols/atg.c]: I'm sorry for my carelessness, I've corrected it

Comment (wssxsxxsx) [reply src/lib/protocols/atg.c]: 4BBCFC3F-BA71-4113-8F8C-ED65C6A80471

**Comment by wssxsxxsx** [comment]
@utoni 
I've added detection for the end of the contents.But I find only Function Code meets, that's why I use the for statement

**Comment by utoni** [comment]
Please add the protocol documentation to `doc/protocols.rst`. See #2521 for a more info.

File: src/lib/protocols/atg.c [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+
+  if(packet->tcp != NULL) {
+  
+   if(packet->payload_packet_len >= 8) {
+      u_int16_t atg_port = ntohs(10001);
+
+    if((packet->tcp->source == atg_port) || (packet->tcp->dest == atg_port)) {
+	  if(packet->payload[0] == 0x01 ) {
+	  for(i=1;i<packet->payload_packet_len;i++){
+		if( packet->payload[i-1]==0x0d && packet->payload[i]==0x0a){
```
Comment (0xA50C1A1) [inline src/lib/protocols/atg.c]: I guess you can replace it with just `if (memcmp(&packet->payload[packet->payload_packet_len-2], "\r\n", 2) == 0)`

**Comment by IvanNardi** [comment]
@wssxsxxsx, as you can see from CI errors, you also need to update also all the (other) unit tests results. Something like:
```
NDPI_FORCE_UPDATING_UTESTS_RESULTS=1 ./tests/do.sh
#  git add all changed files
```

You also need to rebase, please

**Pull request closed without merging**